File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
public:
//     int lps_length(int startIdx,int endIdx,string s){
//         if(startIdx==endIdx)return 1;
//         if(startIdx>endIdx)return 0;
//         if(vect[startIdx][endIdx])return vect[startIdx][endIdx];
//         return vect[startIdx][endIdx]=s[startIdx]==s[endIdx]?2+lps_length(startIdx+1,endIdx-1,s): max(lps_length(startIdx+1,endIdx,s),lps_length(startIdx,endIdx-1,s));
            
//     }
    
    int longestPalindromeSubseq(string s) {
        
        // dp.resize(s.size(),vector<int>(s.size()-1));
        // return lps_length(0,s.size()-1,s);
        int n=s.size();
         vector<vector<int>> dp(n+1,vector<int>(n));
        for(int i=0;i<n;i++) dp[1][i]=1;
        for(int i=2;i<=n;i++) 
            for(int j=0;j<n-i+1;j++) { 
                dp[i][j] = s[j]==s[i+j-1]?2+dp[i-2][j+1]:max(dp[i-1][j],dp[i-1][j+1]);
            }
                return dp[n][0];
    }
};

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.java
class Solution {
    int [][]dp;
//     public int lps(String s,int i,int j){
//         if(i>j)return 0;
//         if(i==j)return 1;
//         if(s.charAt(i)==s.charAt(j))return 2+lps(s,i+1,j-1);
//         if(dp[i][j]!=-1)return dp[i][j];
        
//         return dp[i][j]=Math.max(lps(s,i+1,j),lps(s,i,j-1));
        
//     }
    
    
    public int longestPalindromeSubseq(String s) {
        int n=s.length();
        dp=new int[n+1][n+1];
        // for(int []i:dp)
        //     Arrays.fill(i,-1);
        // return lps(s,0,s.length()-1);
        
        for(int i=1;i<=n;i++){
            for(int j=1;j<=n;j++){
                if(s.charAt(i-1)==s.charAt(n-j))
                    dp[i][j]=1+dp[i-1][j-1];
                else 
                    dp[i][j]=Math.max(dp[i-1][j],dp[i][j-1]);
            }
        } 
        return dp[n][n];
    }
}